src/Data/Patch/MapWithPatchingMove.hs

```diff
@@ -0,0 +1,378 @@
+
+-- | Wrap a @'Map' k (NodeInfo k v)@ representing patch changes into a @'PatchMapWithPatchingMove' k v@, without checking any invariants.
+--
+-- __Warning:__ when using this function, you must ensure that the invariants of 'PatchMapWithPatchingMove' are preserved; they will not be checked.
+unsafePatchMapWithPatchingMove
+  :: Map k (NodeInfo k p) -> PatchMapWithPatchingMove k p
+unsafePatchMapWithPatchingMove = PatchMapWithPatchingMove
+
+-- | Apply the insertions, deletions, and moves to a given 'Map'
+instance (Ord k, Patch p) => Patch (PatchMapWithPatchingMove k p) where
+  type PatchTarget (PatchMapWithPatchingMove k p) = Map k (PatchTarget p)
+  -- TODO: return Nothing sometimes
+  -- Note: the strict application here is critical to ensuring that incremental
+  -- merges don't hold onto all their prerequisite events forever; can we make
+  -- this more robust?
+  apply (PatchMapWithPatchingMove m) old = Just $! insertions `Map.union` (old `Map.difference` deletions)
+    where insertions = flip Map.mapMaybeWithKey m $ \_ ni -> case _nodeInfo_from ni of
+            From_Insert v -> Just v
+            From_Move k p -> applyAlways p <$> Map.lookup k old
+            From_Delete -> Nothing
+          deletions = flip Map.mapMaybeWithKey m $ \_ ni -> case _nodeInfo_from ni of
+            From_Delete -> Just ()
+            _ -> Nothing
+
+-- | Returns all the new elements that will be added to the 'Map'
+patchMapWithPatchingMoveNewElements
+  :: PatchMapWithPatchingMove k p -> [PatchTarget p]
+patchMapWithPatchingMoveNewElements = Map.elems . patchMapWithPatchingMoveNewElementsMap
+
+-- | Return a @'Map' k v@ with all the inserts/updates from the given @'PatchMapWithPatchingMove' k v@.
+patchMapWithPatchingMoveNewElementsMap
+  :: PatchMapWithPatchingMove k p -> Map k (PatchTarget p)
+patchMapWithPatchingMoveNewElementsMap (PatchMapWithPatchingMove p) = Map.mapMaybe f p
+  where f ni = case _nodeInfo_from ni of
+          From_Insert v -> Just v
+          From_Move _ _ -> Nothing
+          From_Delete -> Nothing
+
+-- | Create a 'PatchMapWithPatchingMove' that, if applied to the given 'Map', will sort
+-- its values using the given ordering function.  The set keys of the 'Map' is
+-- not changed.
+patchThatSortsMapWith
+  :: (Ord k, Monoid p)
+  => (PatchTarget p -> PatchTarget p -> Ordering)
+  -> Map k (PatchTarget p) -> PatchMapWithPatchingMove k p
+patchThatSortsMapWith cmp m = PatchMapWithPatchingMove $ Map.fromList $ catMaybes $ zipWith g unsorted sorted
+  where unsorted = Map.toList m
+        sorted = sortBy (cmp `on` snd) unsorted
+        f (to, _) (from, _) = if to == from then Nothing else
+          Just (from, to)
+        reverseMapping = Map.fromList $ catMaybes $ zipWith f unsorted sorted
+        g (to, _) (from, _) = if to == from then Nothing else
+          let Just movingTo = Map.lookup from reverseMapping
+          in Just (to, NodeInfo (From_Move from mempty) $ Just movingTo)
```

Comment:
where does this generate the "To" nodes?  afaict it doesn't ...

Reply:
nevermind, i'm blind; it would be harder to miss this if To were not a type synonym tho ...

Reply:
Yes I had the same thought, and made it its own datatype in the other PR.